### PR TITLE
test clean: test-resources attemptChangeSize tests

### PR DESCRIPTION
**Pull request by zhouyx**
#4145

### CODE REVIEW

File: test/functional/test-resources.js [hunk 1]
```diff
@@ -1058,40 +1046,6 @@ describe('Resources changeSize', () => {
...
-          height: 50};
-      resources.scheduleChangeSize_(resource1, 111, 222, false);
-      resources.mutateWork_();
-      expect(resources.requestsChangeSize_.length).to.equal(0);
-      expect(resource1.changeSize.callCount).to.equal(1);
-      expect(overflowCallbackSpy.callCount).to.equal(1);
-      expect(overflowCallbackSpy.firstCall.args[0]).to.equal(false);
-    });
-
-    it('should NOT change size when in the middle of the viewport', () => {
```
Comment (zhouyx) [inline test/functional/test-resources.js]: duplicate test

File: test/functional/test-resources.js [hunk 2]
```diff
@@ -1058,40 +1046,6 @@ describe('Resources changeSize', () => {
...
-      expect(resource1.changeSize.callCount).to.equal(0);
-      expect(overflowCallbackSpy.callCount).to.equal(1);
-      expect(overflowCallbackSpy.firstCall.args[0]).to.equal(true);
-      expect(overflowCallbackSpy.firstCall.args[1]).to.equal(111);
-      expect(overflowCallbackSpy.firstCall.args[2]).to.equal(222);
-      expect(resource1.getPendingChangeSize().height).to.equal(111);
-      expect(resource1.getPendingChangeSize().width).to.equal(222);
-    });
-
-    it('should NOT change size when below viewport, but decreases', () => {
```
Comment (zhouyx) [inline test/functional/test-resources.js]: test for diff=0, move to top

File: test/functional/test-resources.js (hunk 1 above)
Comment (sriramkrish85) [inline test/functional/test-resources.js]: why is the next it statement also being deleted?

File: test/functional/test-resources.js (hunk 2 above)
Comment (sriramkrish85) [inline test/functional/test-resources.js]: can you paste the links of the duplicate tests?

File: test/functional/test-resources.js (hunk 1 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: https://github.com/zhouyx/amphtml/blob/3abb80f69c45c9beae069675350e10b4697069ad/test/functional/test-resources.js#L1125

File: test/functional/test-resources.js (hunk 2 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: https://github.com/zhouyx/amphtml/blob/3abb80f69c45c9beae069675350e10b4697069ad/test/functional/test-resources.js#L998

File: test/functional/test-resources.js [hunk 3]
```diff
@@ -994,23 +994,11 @@ describe('Resources changeSize', () => {
       vsyncSpy = sandbox.stub(resources.vsync_, 'run');
     });
 
-    it('should NOT change size and calls overflowCallback', () => {
```
Comment (zhouyx) [inline test/functional/test-resources.js]: https://github.com/zhouyx/amphtml/blob/3abb80f69c45c9beae069675350e10b4697069ad/test/functional/test-resources.js#L1125

File: test/functional/test-resources.js [hunk 4]
```diff
@@ -994,23 +994,11 @@ describe('Resources changeSize', () => {
...
-      expect(resource1.changeSize.callCount).to.equal(0);
-      expect(overflowCallbackSpy.callCount).to.equal(1);
-      expect(overflowCallbackSpy.firstCall.args[0]).to.equal(true);
-      expect(overflowCallbackSpy.firstCall.args[1]).to.equal(111);
-      expect(overflowCallbackSpy.firstCall.args[2]).to.equal(222);
-      expect(resource1.getPendingChangeSize().height).to.equal(111);
-      expect(resource1.getPendingChangeSize().width).to.equal(222);
-    });
-
-    it('should change size when new height/width is lower', () => {
```
Comment (zhouyx) [inline test/functional/test-resources.js]: https://github.com/zhouyx/amphtml/blob/3abb80f69c45c9beae069675350e10b4697069ad/test/functional/test-resources.js#L1117

File: test/functional/test-resources.js [hunk 5]
```diff
@@ -1058,40 +1046,6 @@ describe('Resources changeSize', () => {
       expect(overflowCallbackSpy.firstCall.args[0]).to.equal(false);
     });
 
-    it('should change size when slightly above the viewport', () => {
```
Comment (zhouyx) [inline test/functional/test-resources.js]: This test actually tests element resize when below viewport. Duplicate of
https://github.com/zhouyx/amphtml/blob/3abb80f69c45c9beae069675350e10b4697069ad/test/functional/test-resources.js#L1038

File: test/functional/test-resources.js [hunk 6]
```diff
@@ -1160,6 +1114,27 @@ describe('Resources changeSize', () => {
...
+
+    it('in viewport should NOT change size and calls overflowCallback', () => {
+      resources.scheduleChangeSize_(resource1, 111, 222, false);
+      resources.mutateWork_();
+      expect(resources.requestsChangeSize_.length).to.equal(0);
+      expect(resource1.changeSize.callCount).to.equal(0);
+      expect(overflowCallbackSpy.callCount).to.equal(1);
+      expect(overflowCallbackSpy.firstCall.args[0]).to.equal(true);
+      expect(overflowCallbackSpy.firstCall.args[1]).to.equal(111);
+      expect(overflowCallbackSpy.firstCall.args[2]).to.equal(222);
```
Comment (lannka) [inline test/functional/test-resources.js]: you can merge the 3 lines to
`expect(overflowCallbackSpy.firstCall.args).to.jsonEqual([true, 111, 222]);`

File: test/functional/test-resources.js (hunk 4 above)
Comment (sriramkrish85) [inline test/functional/test-resources.js]: This test is the opposite of what is being deleted?
`in vp should NOT call overflowCallback if new height smaller`

File: test/functional/test-resources.js [hunk 7]
```diff
@@ -1160,6 +1114,27 @@ describe('Resources changeSize', () => {
...
+      expect(resource1.getPendingChangeSize().height).to.equal(111);
+      expect(resource1.getPendingChangeSize().width).to.equal(222);
```
Comment (lannka) [inline test/functional/test-resources.js]: Same here. jsonEqual

File: test/functional/test-resources.js [hunk 8]
```diff
@@ -1160,6 +1114,27 @@ describe('Resources changeSize', () => {
       expect(resources.relayoutTop_).to.equal(resource1.layoutBox_.top);
     });
 
+    it('in vp should NOT call overflowCallback if new height smaller', () => {
+      resources.scheduleChangeSize_(resource1, 10, 11, false);
+      resources.mutateWork_();
+      expect(resources.requestsChangeSize_.length).to.equal(0);
+      expect(resource1.changeSize.callCount).to.equal(0);
```
Comment (lannka) [inline test/functional/test-resources.js]: `assert(resource1.changeSize.notCalled);`

File: test/functional/test-resources.js [hunk 9]
```diff
@@ -1160,6 +1114,27 @@ describe('Resources changeSize', () => {
...
+      expect(overflowCallbackSpy.callCount).to.equal(0);
```
Comment (lannka) [inline test/functional/test-resources.js]: Same here

File: test/functional/test-resources.js [hunk 10]
```diff
@@ -1160,6 +1114,27 @@ describe('Resources changeSize', () => {
       expect(resources.relayoutTop_).to.equal(resource1.layoutBox_.top);
     });
 
+    it('in vp should NOT call overflowCallback if new height smaller', () => {
+      resources.scheduleChangeSize_(resource1, 10, 11, false);
+      resources.mutateWork_();
+      expect(resources.requestsChangeSize_.length).to.equal(0);
```
Comment (lannka) [inline test/functional/test-resources.js]: here can use `.to.be.empty`

File: test/functional/test-resources.js [hunk 11]
```diff
@@ -1160,6 +1114,27 @@ describe('Resources changeSize', () => {
...
+      expect(resource1.changeSize.callCount).to.equal(0);
+      expect(overflowCallbackSpy.callCount).to.equal(0);
+    });
+
+    it('in viewport should NOT change size and calls overflowCallback', () => {
+      resources.scheduleChangeSize_(resource1, 111, 222, false);
+      resources.mutateWork_();
+      expect(resources.requestsChangeSize_.length).to.equal(0);
+      expect(resource1.changeSize.callCount).to.equal(0);
+      expect(overflowCallbackSpy.callCount).to.equal(1);
```
Comment (lannka) [inline test/functional/test-resources.js]: `assert(callback.calledOnce);`

File: test/functional/test-resources.js (hunk 4 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: But if you take a closer look at the code, this test do `expect(resource1.changeSize.callCount).to.equal(0);`. The description is misleading

Comment (sriramkrish85) [inline test/functional/test-resources.js]: Then you should fix the tests to cater to these scenarios as well

**Comment by lannka** [comment]
Please try to optimize the asserts with those various method provided by [Chai](http://chaijs.com/api/bdd/) and [Sinon](http://sinonjs.org/). It's not just simplify code, more importantly it gives better error messages.

For example:
`expect(array.length).to.equal(0)` prints error message: `1 does not equal to 0`
`expect(array).to.be.empty` prints error message: `Array [ 'item' ] is not empty`

File: test/functional/test-resources.js (hunk 4 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: From the code of the test, I suppose the old test was intended to check if `overflowCallback` not getting called when height decrease (which is a special case) but ended up with a wrong name.
I don't see why this scenario `'should change size when new height/width is lower'` is special and needed to be tested. Basically if the element is in viewport we will NOT change change size and NOT call `oveflowCallback` (what the new test cover), if the element is outside viewport, we don't care if the new height is smaller or not (we have test about element above/below viewport).

File: test/functional/test-resources.js (hunk 6 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: done

File: test/functional/test-resources.js (hunk 8 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: use `expect(resource1.changeSize).to.not.been.called` instead.

File: test/functional/test-resources.js (hunk 10 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: done

File: test/functional/test-resources.js (hunk 11 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: used `expect(overflowCallbackSpy).to.be.calledOnce`;

**Comment by zhouyx** [comment]
@lannka I changed the test according to what you suggest. PTAL

**Comment by zhouyx** [comment]
@sriramkrish85 @lannka  PTAL

File: test/functional/test-resources.js [hunk 12]
```diff
@@ -994,33 +994,21 @@ describe('Resources changeSize', () => {
...
+      expect(overflowCallbackSpy).to.not.been.called;
     });
 
     it('should change size when forced', () => {
       resources.scheduleChangeSize_(resource1, 111, 222, true);
       resources.mutateWork_();
-      expect(resources.requestsChangeSize_.length).to.equal(0);
-      expect(resource1.changeSize.callCount).to.equal(1);
-      expect(overflowCallbackSpy.callCount).to.equal(1);
+      expect(resources.requestsChangeSize_.length).to.be.empty;
```
Comment (lannka) [inline test/functional/test-resources.js]: remove `.length`

File: test/functional/test-resources.js [hunk 13]
```diff
@@ -994,33 +994,21 @@ describe('Resources changeSize', () => {
...
+      expect(resource1.changeSize).to.be.calledOnce;
```
Comment (lannka) [inline test/functional/test-resources.js]: Oh, this is working? It's great then. It means we have the sinon support in chai.

I was using `sinon.assert.calledOnce(spy)`.

File: test/functional/test-resources.js [hunk 14]
```diff
@@ -1031,19 +1019,19 @@ describe('Resources changeSize', () => {
...
     });
 
     it('should change size when active', () => {
       resource1.element.contains = () => true;
       resources.scheduleChangeSize_(resource1, 111, 222, false);
       resources.mutateWork_();
-      expect(resources.requestsChangeSize_.length).to.equal(0);
-      expect(resource1.changeSize.callCount).to.equal(1);
-      expect(overflowCallbackSpy.callCount).to.equal(1);
+      expect(resources.requestsChangeSize_.length).to.be.empty;
```
Comment (lannka) [inline test/functional/test-resources.js]: same here

File: test/functional/test-resources.js [hunk 15]
```diff
@@ -1052,46 +1040,12 @@ describe('Resources changeSize', () => {
           height: 50};
       resources.scheduleChangeSize_(resource1, 111, 222, false);
       resources.mutateWork_();
-      expect(resources.requestsChangeSize_.length).to.equal(0);
-      expect(resource1.changeSize.callCount).to.equal(1);
-      expect(overflowCallbackSpy.callCount).to.equal(1);
+      expect(resources.requestsChangeSize_.length).to.be.empty;
```
Comment (lannka) [inline test/functional/test-resources.js]: same here

**Comment by zhouyx** [comment]
done, @lannka PTAL

**Comment by sriramkrish85** [comment]
LGTM. wait for @lannka to LGTM

File: test/functional/test-resources.js [hunk 16]
```diff
@@ -1173,20 +1144,19 @@ describe('Resources changeSize', () => {
...
 
       resources.mutateWork_();
-      expect(resources.requestsChangeSize_.length).to.equal(0);
-      expect(resource1.changeSize.callCount).to.equal(1);
-      expect(resource1.changeSize.firstCall.args[0]).to.equal(111);
-      expect(resource1.changeSize.firstCall.args[1]).to.equal(222);
-      expect(overflowCallbackSpy.callCount).to.equal(2);
+      expect(resources.requestsChangeSize_).to.be.empty;
+      expect(resource1.changeSize).to.be.calledOnce;
+      expect(resource1.changeSize.firstCall.args).to.jsonEqual([111, 222]);
```
Comment (lannka) [inline test/functional/test-resources.js]: I'm not sure about this, but could you try if chai-sinon support syntax like `to.be.calledWith()`, or something similar?

**Comment by lannka** [comment]
LGTM

File: test/functional/test-resources.js (hunk 16 above)
Comment (zhouyx) [inline test/functional/test-resources.js]: It is actually working, do you want me to change to that format?

Comment (zhouyx) [inline test/functional/test-resources.js]: nvm, already changed

**Pull request merged**